crates/core/src/delta_datafusion/table_provider.rs

```diff
@@ -570,12 +571,87 @@ impl<'a> DeltaScanBuilder<'a> {
+            let mut action_with_normalized_path = action.clone();
+            let (normalized_path, needs_bucket_root) = crate::logstore::normalize_add_path_for_scan(
+                &root,
+                action_with_normalized_path.path.as_str(),
+            );
+            action_with_normalized_path.path = normalized_path;
+            need_bucket_root_store |= needs_bucket_root;
+
+            // Build PartitionedFile with ObjectMeta that matches the selected store
+            let ts_secs = action_with_normalized_path.modification_time / 1000;
+            let ts_ns = (action_with_normalized_path.modification_time % 1000) * 1_000_000;
+            let last_modified = Utc.from_utc_datetime(
+                &DateTime::from_timestamp(ts_secs, ts_ns as u32)
+                    .expect(
+                        "Invalid timestamp in PartitionedFile: modification_time is out of range",
+                    )
+                    .naive_utc(),
+            );
+
+            let object_meta = if root.scheme() == "file" && allow_unrestricted_file_access {
+                // Use a root-scoped filesystem store; provide absolute path without leading '/'
+                ObjectMeta {
+                    location: crate::logstore::object_store_path_for_file_root(
+                        &root,
+                        action_with_normalized_path.path.as_str(),
+                    ),
+                    last_modified,
+                    size: action_with_normalized_path.size as u64,
+                    e_tag: None,
+                    version: None,
+                }
+            } else {
+                ObjectMeta {
+                    location: action_with_normalized_path.path.clone().into(),
+                    last_modified,
+                    size: action_with_normalized_path.size as u64,
+                    e_tag: action_with_normalized_path
+                        .tags
+                        .as_ref()
+                        .and_then(|tags| tags.get("etag"))
+                        .cloned()
+                        .flatten(),
+                    version: None,
+                }
+            };
+
+            // Build PartitionedFile from action to ensure partition values align with schema
+            // and DataFusion expectations
+            let mut part = partitioned_file_from_action(
+                &action_with_normalized_path,
+                table_partition_cols,
+                &schema,
+            );
+            // Overwrite object_meta with the one computed above to preserve normalized path
+            part.object_meta = object_meta;
 
+            // Optionally append the virtual file column as an extra partition value
             if config.file_column_name.is_some() {
                 let partition_value = if config.wrap_partition_values {
                     wrap_partition_value_in_dict(ScalarValue::Utf8(Some(action.path.clone())))
```

Comment:
This uses the path from the original action rather than the normalized one, is that intended?